File: android/data/src/main/kotlin/mesh/data/Entities.kt
package mesh.data

import androidx.room.Entity
import androidx.room.PrimaryKey
import mesh.core.protocol.MessageStatus

@Entity(tableName = "contacts")
data class ContactEntity(
    @PrimaryKey val nodeId: String, // Hex string
    val identityKey: ByteArray,
    val verified: Boolean
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is ContactEntity) return false
        return nodeId == other.nodeId
    }

    override fun hashCode(): Int = nodeId.hashCode()
}

@Entity(tableName = "sessions")
data class SessionEntity(
    @PrimaryKey val nodeId: String, // Hex string of peer NodeId
    val serializedState: ByteArray // Serliazed ratchet state BLOB
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is SessionEntity) return false
        return nodeId == other.nodeId
    }

    override fun hashCode(): Int = nodeId.hashCode()
}

@Entity(tableName = "messages")
data class MessageEntity(
    @PrimaryKey val id: String,
    val convoId: String, // Conversation group or contact NodeID
    val isSent: Boolean, // True = outbound, False = inbound
    val status: String, // MessageStatus string (PENDING, SENT, DELIVERED, READ, FAILED)
    val timestamp: Long, // Epoch ms
    val ciphertext: ByteArray
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is MessageEntity) return false
        return id == other.id
    }

    override fun hashCode(): Int = id.hashCode()
}

@Entity(tableName = "outbox")
data class OutboxEntity(
    @PrimaryKey val packetId: String,
    val destNodeId: String,
    val status: String, // PENDING, IN_FLIGHT, SENT, FAILED
    val data: ByteArray, // Encoded MeshPacket binary frame
    val retryCount: Int = 0,
    val lastRetryTime: Long = 0L
) {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is OutboxEntity) return false
        return packetId == other.packetId
    }

    override fun hashCode(): Int = packetId.hashCode()
}

File: ios/Sources/DataStore/Entities.swift
import Foundation
import GRDB
import MeshCore

public struct ContactRecord: Codable, FetchableRecord, MutablePersistableRecord, TableRecord {
    public static let databaseTableName = "contacts"
    
    public var nodeId: String // Hex
    public var identityKey: Data
    public var verified: Bool
    
    public init(nodeId: String, identityKey: Data, verified: Bool) {
        self.nodeId = nodeId
        self.identityKey = identityKey
        self.verified = verified
    }
}

public struct SessionRecord: Codable, FetchableRecord, MutablePersistableRecord, TableRecord {
    public static let databaseTableName = "sessions"
    
    public var nodeId: String // Hex of peer NodeID
    public var serializedState: Data
    
    public init(nodeId: String, serializedState: Data) {
        self.nodeId = nodeId
        self.serializedState = serializedState
    }
}

public struct MessageRecord: Codable, FetchableRecord, MutablePersistableRecord, TableRecord {
    public static let databaseTableName = "messages"
    
    public var id: String
    public var convoId: String
    public var isSent: Bool
    public var status: String
    public var timestamp: Int64 // Epoch ms
    public var ciphertext: Data
    
    public init(id: String, convoId: String, isSent: Bool, status: String, timestamp: Int64, ciphertext: Data) {
        self.id = id
        self.convoId = convoId
        self.isSent = isSent
        self.status = status
        self.timestamp = timestamp
        self.ciphertext = ciphertext
    }
}

public struct OutboxRecord: Codable, FetchableRecord, MutablePersistableRecord, TableRecord {
    public static let databaseTableName = "outbox"
    
    public var packetId: String
    public var destNodeId: String
    public var status: String
    public var data: Data
    public var retryCount: Int
    public var lastRetryTime: Int64
    
    public init(packetId: String, destNodeId: String, status: String, data: Data, retryCount: Int = 0, lastRetryTime: Int64 = 0) {
        self.packetId = packetId
        self.destNodeId = destNodeId
        self.status = status
        self.data = data
        self.retryCount = retryCount
        self.lastRetryTime = lastRetryTime
    }
}
